### ISSUE TITLE
Unable to retrieve private post/menu using wp-graphql-jwt-authentication/Bearer Token

### ISSUE THREAD

**Issue by kpratik2015** [wp-graphql/wp-graphql#1487]
Hi so I've been struggling with this issue for last 3 days and I've reached a dead end. This issue started happening after I upgraded my wordpress to 5.5.1 from 5.5. So last I checked, 3-4 weeks back the setup I had was working perfectly. What still works right now is `loginWithCookies` but it does not retrieve me menu which is `private` as there's no parent.
Anyway, I want to make `authToken` generated by `login` mutation work.

What I've checked or done so far:

**Install latest release of each plugin**
- WP GraphQL (0.13.1)
- WPGraphQL CORS (1.1.0)
- WPGraphQL JWT Authentication (0.4.1)
- Advanced Custom Fields (5.9.1)
- Custom Post Type UI (1.8.1)
- Gutenberg (9.1.0)
- Wordfence Security (7.4.11)

**`.htaccess` various modification attempts**
```Apache Conf
SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
SetEnvIf Origin "^http(s)?://(.+\.)?(real-domain.com|localhost)(:[0-9]+)?[/]?$" origin_is=$0 
Header always set Access-Control-Allow-Origin %{origin_is}e env=origin_is
# Header always set Access-Control-Allow-Headers 'Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With, cache-control' env=origin_is
# Header always set Access-Control-Allow-Credentials 'true' env=origin_is
```
I found that via Postman, if I put `Origin: http://localhost:3000` in request header then only in response `Access-Control-Allow-Origin`, I get `http://localhost:3000` instead of `*`

**WPGraphQL CORS settings**

![image](https://user-images.githubusercontent.com/14140930/95123253-47a9b280-076f-11eb-8162-88a7110cef66.png)

**Exploring postman**

This is request/response with `login mutation`

![image](https://user-images.githubusercontent.com/14140930/95123668-ed5d2180-076f-11eb-8d7c-8f52b3351f92.png)

Now this is request/response for posts with status `PRIVATE`. Note: Bearer token is of admin user. Also tried Editor user.

![image](https://user-images.githubusercontent.com/14140930/95123878-41680600-0770-11eb-8a47-8ed88b83b527.png)

There is one `PRIVATE` post which I do get when I use `loginWithCookies`. Also, is `X-JWT-Auth` suppose to have different token than the Bearer token sent? I've also tried `viewer` query and it just gives me null. If I put incorrect or expired token then I get expected errors (so `Authorization` is going through). I do get `PUBLIC` status posts, which anyway is possible without Bearer token.

This is not limited to localhost too. I've deployed my front-end on same domain - albeit different sub-domain and it gives same results. 

Any help or pointing in right direction is appreciated.

**Comment by kpratik2015** [comment]
@jasonbahl Hi, I know you may not have the time but could you offer any insight or way for me to debug this on my own? Perhaps wordpress side?

**Comment by kpratik2015** [comment]
I added a lot of logs at many places and have got a lead on `Request.php` which outputs `$wp_rest_auth_cookie` as malformed. I guess I will try to resolve the issue myself without knowing PHP :/

**Comment by jasonbahl** [comment]
@kpratik2015 hey, I'm not really sure what the issue is. 

Can you get private posts when using GraphiQL in the WordPress dashboard? Or are you not able to get private posts there too?

**Comment by kpratik2015** [comment]
> @kpratik2015 hey, I'm not really sure what the issue is.
> 
> Can you get private posts when using GraphiQL in the WordPress dashboard? Or are you not able to get private posts there too?

Hi Jason! Glad to have grabbed your attention. I am able to get private posts via GraphiQL. But somehow using postman with Bearer Token I am not able to get.`is_user_logged_in()` is false when I query, even though it does becomes true when I generate Bearer Token via `login` mutation.

**Comment by jasonbahl** [comment]
Ok, so the fact that you can get private posts in GraphiQL in the admin leads me to believe WPGraphQL is working properly and it's something else. Likely one of the extensions. 🤔

**Comment by kpratik2015** [comment]
> Ok, so the fact that you can get private posts in GraphiQL in the admin leads me to believe WPGraphQL is working properly and it's something else. Likely one of the extensions. 🤔

Could it be wp-graphql-jwt-authentication?

Sorry I can't help a lot as I don't know PHP. I am just trying to print into debug.log via debug mode of wordpress.

The `login` mutation does works though as I do get this:

```
[09-Oct-2020 21:42:22 UTC] Array
(
    [0] => current user
    [1] => WP_User Object
        (
            [data] => stdClass Object
                (
                    [ID] => 4
                    [user_login] => api
                    [user_pass] => *******
                    [user_nicename] => api
                    [user_email] => api@dummy.com
                    [user_url] => 
                    [user_registered] => 2020-10-04 14:15:10
                    [user_activation_key] => 
                    [user_status] => 0
                    [display_name] => API
                )

            [ID] => 4
            [caps] => Array
                (
                    [editor] => 1
                )

            [cap_key] => wpdb_capabilities
            [roles] => Array
                (
                    [0] => editor
                )

            [allcaps] => Array
                (
                    [moderate_comments] => 1
                    [manage_categories] => 1
                    [manage_links] => 1
                    [upload_files] => 1
                    [unfiltered_html] => 1
                    [edit_posts] => 1
                    [edit_others_posts] => 1
                    [edit_published_posts] => 1
                    [publish_posts] => 1
                    [edit_pages] => 1
                    [read] => 1
                    [level_7] => 1
                    [level_6] => 1
                    [level_5] => 1
                    [level_4] => 1
                    [level_3] => 1
                    [level_2] => 1
                    [level_1] => 1
                    [level_0] => 1
                    [edit_others_pages] => 1
                    [edit_published_pages] => 1
                    [publish_pages] => 1
                    [delete_pages] => 1
                    [delete_others_pages] => 1
                    [delete_published_pages] => 1
                    [delete_posts] => 1
                    [delete_others_posts] => 1
                    [delete_published_posts] => 1
                    [delete_private_posts] => 1
                    [edit_private_posts] => 1
                    [read_private_posts] => 1
                    [delete_private_pages] => 1
                    [edit_private_pages] => 1
                    [read_private_pages] => 1
                    [edit_blocks] => 1
                    [edit_others_blocks] => 1
                    [publish_blocks] => 1
                    [read_private_blocks] => 1
                    [read_blocks] => 1
                    [delete_blocks] => 1
                    [delete_private_blocks] => 1
                    [delete_published_blocks] => 1
                    [delete_others_blocks] => 1
                    [edit_private_blocks] => 1
                    [edit_published_blocks] => 1
                    [create_blocks] => 1
                    [wpseo_bulk_edit] => 1
                    [wpseo_manage_redirects] => 1
                    [activate_plugins] => 1
                    [editor] => 1
                )

            [filter] => 
            [site_id:WP_User:private] => 1
        )

)
```

But on executing query, I get:

```
(
    [0] => current user
    [1] => WP_User Object
        (
            [data] => stdClass Object
                (
                )

            [ID] => 0
            [caps] => Array
                (
                )

            [cap_key] => 
            [roles] => Array
                (
                )

            [allcaps] => Array
                (
                )

            [filter] => 
            [site_id:WP_User:private] => 0
        )

)
```
Wrote this log in `get_app_context()` function of `wp-graphql.php`. Is there any other useful place to put log and check?

**Comment by jasonbahl** [comment]
Ok, so I'm not able to replicate this. 

I'm on: 

- WordPress 5.5.1
- WPGraphQL v0.13.3
- WPGraphQL v0.4.1

I created a private post, id `1634`: 

![Screen Shot 2020-10-09 at 3 46 40 PM](https://user-images.githubusercontent.com/1260765/95634229-a0ff3380-0a46-11eb-8d07-6cfaa7f753c7.png)

If I query this post as a public user, I get null, as expected:

![Screen Shot 2020-10-09 at 3 47 26 PM](https://user-images.githubusercontent.com/1260765/95634275-bc6a3e80-0a46-11eb-95ca-030b4e388ff5.png)

If I use the login mutation (provided from WPGraphQL JWT Auth)

![Screen Shot 2020-10-09 at 3 48 10 PM](https://user-images.githubusercontent.com/1260765/95634318-d7d54980-0a46-11eb-9d30-cc16f7e345c6.png)

Then take the `authToken` that was returned from the login mutation and use it in the header to make the same query for the Post by ID (I also included `viewer { username }` to show that the token is indeed setting the current user as `jasonbahl`, the same user that I logged in as):

![Screen Shot 2020-10-09 at 3 49 18 PM](https://user-images.githubusercontent.com/1260765/95634375-ffc4ad00-0a46-11eb-8c1f-eaa87aa0c43f.png)

**Comment by kpratik2015** [comment]
Okay, probably something unique in my setup then. Could you let me know where it might be helpful to put logs to check?

Response for me with above query:

![image](https://user-images.githubusercontent.com/14140930/95635206-637fc200-0aa9-11eb-952d-32b2453ac8d8.png)

**Comment by kpratik2015** [comment]
So turns out the culprit was Jetpack plugin (Version 9.0.2). On deactivating it, I was able to get private posts. Some googling helped me to nail on this path. Particularly, [this](https://wordpress.org/support/topic/jetpack-and-rest-api/).

However, there's still one issue I have that I am unable to retrieve menu with no parent i.e. private menu for Editor role user. Admin user can get it and I believe an Editor user should as well since it can access private post, maybe?

Thanks @jasonbahl  :)

**Comment by kpratik2015** [comment]
Closing this issue as I was able to edit source code (I know it's not optimum) directly to meet my condition.

I basically wanted private data to be visible for a particular user. The [Members](https://wordpress.org/plugins/members/) plugin helped me partly to create a custom role that only had read granted for everything. Even then with the new role, my user wasn't able to get details of a private post (private menu still accessible only by admin anyway). So my last resort was to just put an `if` condition in `Model.php` file of plugin inside `get_visibility()` function.

```php
if( $this->current_user && $this->current_user->data->user_login === 'xxx' ) {
   $this->visibility = 'public';
}
```

Posting my solution for helping someone. Also do add Jetpack as a potential conflicting plugin to use with this.
